File: examples/client-gen/tictactoe/types/sign.py
from __future__ import annotations
import typing
from dataclasses import dataclass
from anchorpy.borsh_extension import EnumForCodegen
import borsh_construct as borsh


class XJSON(typing.TypedDict):
    kind: typing.Literal["X"]


class OJSON(typing.TypedDict):
    kind: typing.Literal["O"]


@dataclass
class X:
    discriminator: typing.ClassVar = 0
    kind: typing.ClassVar = "X"

    @classmethod
    def to_json(cls) -> XJSON:
        return XJSON(
            kind="X",
        )

    @classmethod
    def to_encodable(cls) -> dict:
        return {
            "X": {},
        }


@dataclass
class O:
    discriminator: typing.ClassVar = 1
    kind: typing.ClassVar = "O"

    @classmethod
    def to_json(cls) -> OJSON:
        return OJSON(
            kind="O",
        )

    @classmethod
    def to_encodable(cls) -> dict:
        return {
            "O": {},
        }


SignKind = typing.Union[X, O]
SignJSON = typing.Union[XJSON, OJSON]


def from_decoded(obj: dict) -> SignKind:
    if not isinstance(obj, dict):
        raise ValueError("Invalid enum object")
    if "X" in obj:
        return X()
    if "O" in obj:
        return O()
    raise ValueError("Invalid enum object")


def from_json(obj: SignJSON) -> SignKind:
    if obj["kind"] == "X":
        return X()
    if obj["kind"] == "O":
        return O()
    kind = obj["kind"]
    raise ValueError(f"Unrecognized enum kind: {kind}")


layout = EnumForCodegen("X" / borsh.CStruct(), "O" / borsh.CStruct())

File: ts-reference/examples/tic-tac-toe/generated-client/types/Sign.ts
import { PublicKey } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh"

export interface XJSON {
  kind: "X"
}

export class X {
  readonly discriminator = 0
  readonly kind = "X"

  toJSON(): XJSON {
    return {
      kind: "X",
    }
  }

  toEncodable() {
    return {
      X: {},
    }
  }
}

export interface OJSON {
  kind: "O"
}

export class O {
  readonly discriminator = 1
  readonly kind = "O"

  toJSON(): OJSON {
    return {
      kind: "O",
    }
  }

  toEncodable() {
    return {
      O: {},
    }
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function fromDecoded(obj: any): types.SignKind {
  if (typeof obj !== "object") {
    throw new Error("Invalid enum object")
  }

  if ("X" in obj) {
    return new X()
  }
  if ("O" in obj) {
    return new O()
  }

  throw new Error("Invalid enum object")
}

export function fromJSON(obj: types.SignJSON): types.SignKind {
  switch (obj.kind) {
    case "X": {
      return new X()
    }
    case "O": {
      return new O()
    }
  }
}

export function layout(property?: string) {
  const ret = borsh.rustEnum([borsh.struct([], "X"), borsh.struct([], "O")])
  if (property !== undefined) {
    return ret.replicate(property)
  }
  return ret
}
